source-snowflake/discovery.go

```diff
@@ -0,0 +1,383 @@
+	"github.com/invopop/jsonschema"
+	"github.com/jmoiron/sqlx"
+	"github.com/sirupsen/logrus"
+	log "github.com/sirupsen/logrus"
+)
+
+func (snowflakeDriver) Discover(ctx context.Context, req *pc.Request_Discover) (*pc.Response_Discovered, error) {
+	var cfg = new(config)
+	if err := pf.UnmarshalStrict(req.ConfigJson, cfg); err != nil {
+		return nil, fmt.Errorf("error parsing config json: %w", err)
+	}
+	cfg.SetDefaults()
+
+	log.Info("performing discovery")
+
+	var db, err = connectSnowflake(ctx, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to snowflake: %w", err)
+	}
+	defer db.Close()
+
+	discoveryResults, err := performSnowflakeDiscovery(ctx, cfg, db)
+	if err != nil {
+		log.WithField("err", err).Fatal("error discovering snowflake collections")
+	}
+
+	bindings, err := catalogFromDiscovery(cfg, discoveryResults)
+	if err != nil {
+		log.WithField("err", err).Fatal("error converting discovery info to bindings")
+	}
+
+	return &pc.Response_Discovered{Bindings: bindings}, nil
+}
+
+type snowflakeDiscoveryResults struct {
+	Tables      []*snowflakeDiscoveryTable
+	Columns     []*snowflakeDiscoveryColumn
+	PrimaryKeys []*snowflakeDiscoveryPrimaryKey
+}
+
+type snowflakeDiscoveryTable struct {
+	Database string `db:"database_name"`
+	Schema   string `db:"schema_name"`
+	Name     string `db:"name"`
+	Kind     string `db:"kind"`
+	Comment  string `db:"comment"`
+}
+
+type snowflakeDiscoveryColumn struct {
+	Database         string  `db:"TABLE_CATALOG"`
+	Schema           string  `db:"TABLE_SCHEMA"`
+	Table            string  `db:"TABLE_NAME"`
+	Name             string  `db:"COLUMN_NAME"`
+	Index            int     `db:"ORDINAL_POSITION"`
+	DataType         string  `db:"DATA_TYPE"`
+	Comment          *string `db:"COMMENT"`
+	IsNullable       string  `db:"IS_NULLABLE"`
+	NumericScale     *int    `db:"NUMERIC_SCALE"`
+	NumericPrecision *int    `db:"NUMERIC_PRECISION"`
+}
+
+type snowflakeDiscoveryPrimaryKey struct {
+	Database   string `db:"database_name"`
+	Schema     string `db:"schema_name"`
+	Table      string `db:"table_name"`
+	ColumnName string `db:"column_name"`
+	KeySeq     int    `db:"key_sequence"`
+}
+
+func performSnowflakeDiscovery(ctx context.Context, cfg *config, db *sql.DB) (*snowflakeDiscoveryResults, error) {
+	log.Debug("performing discovery")
+
+	// We have to use `SHOW FOO` queries to get some of the information we need, such
+	// as primary keys. But there's no simple way to select just the columns we need
+	// from these nonstandard not-a-table queries, and that makes them unstable if
+	// new columns were to be added in the future.
+	//
+	// So we wrap the database connection with SQLX and use its struct-tag reflection
+	// to select the columns of interest *by name*, and set Unsafe() so that it will
+	// ignore unmatched columns.
+	var xdb = sqlx.NewDb(db, "snowflake").Unsafe()
+	var tables []*snowflakeDiscoveryTable
+	if err := xdb.Select(&tables, "SHOW TABLES;"); err != nil {
```

Comment:
This seems to work, but can you not get what is needed from `information_schema.tables`?

Reply:
In this case yeah, I think all the information is present in both `SHOW TABLES` and `information_schema.tables`.

The real sticking point is `SHOW PRIMARY KEYS` which cannot be replaced with an `information_schema` query at all. IIRC, because of that I originally _only_ used the `SHOW FOO` style queries here for consistency, but then it turned out that `SHOW COLUMNS` didn't give us all of the information we need while `information_schema.columns` did, so now we're forced to have an inconsistent mix of data sources.

So this could in principle be swapped out for `SELECT * FROM information_schema.tables` with a little bit of name-changing in the corresponding struct tags, but I'm not convinced either option is strictly better so I'd rather just leave it as-is.